[full ci] Prevent vic-machine from deleting non cVMs

andrewtchin opened the pull request:
Fixes #6603

cgtexmex commented on lib/install/management/delete.go:
```diff
@@ -46,11 +47,36 @@ func (d *Dispatcher) DeleteVCH(conf *config.VirtualContainerHostConfigSpec) erro
 	}
+	if d.parentResourcepool != nil {
+		if d.parentResourcepool.Reference() == defaultrp.Reference() {
+			return fmt.Errorf("Targed VCH is in cluster's default resource pool. Refusing to delete it.")
```
Why would we not remove if it's in the cluster default?  We won't install there, so it's been moved there...why not remove it?

andrewtchin replied:
I thought this was a requirement, but that makes sense to me. I removed this

cgtexmex replied:
Well..hell...looking at the issue I see why you'd think that was a requirement..I think the first item should have said "vic-machine _create_ will fail if the target rp is the default rp"...was really thinking of adding a new validation rule.  I'm not sure that's really needed...

I think we prevent the install into a cluster default RP by creating an RP on each deploy.  Previously there was an issue that would (in some cases) result in a VCH deployed to the default RP, but I think since 1.2.1 that's fixed...

andrewtchin replied:
yea you're right 👍

cgtexmex commented on lib/install/management/delete.go:
```diff
@@ -46,11 +47,36 @@ func (d *Dispatcher) DeleteVCH(conf *config.VirtualContainerHostConfigSpec) erro
+			return err
+		}
+		// Assume the cVMs and RP have already been deleted
+		log.Warnf("Proceeding with delete of VCH due to --force")
```
Will there be later feedback / logging that no containerVMs have been removed?  If not should that be stated here?  i.e. "No container VMs found, but proceeding with delete of VCH due to use of --force"

IMO there's a possibility that the containerVMs exist else where, so they would be "orphaned". Just want to make sure the user understands that no other VMs will be removed...

andrewtchin replied:
Good point. Added this

andrewtchin commented on lib/install/management/delete.go:
```diff
@@ -222,13 +235,29 @@ func (d *Dispatcher) DeleteVCHInstances(vmm *vm.VirtualMachine, conf *config.Vir
 		if ok {
-			// child is vch; detach all attached disks so later removal of images is successful
+			// Do not delete a VCH in the target RP if it is not the target VCH
+			if child.Reference() != vmm.Reference() {
```
This was to address if the user OOB moved a VCH into another VCH's RP. I will open an issue to disable moving a VCH

andrewtchin commented on lib/install/management/delete.go:
```diff
@@ -46,10 +46,28 @@ func (d *Dispatcher) DeleteVCH(conf *config.VirtualContainerHostConfigSpec) erro
-		// if container delete failed, do not remove anything else
-		log.Infof("Specify --force to force delete")
-		return err
+	d.parentResourcepool, err = d.getComputeResource(vmm, conf)
```
This and the next change below was to allow a VCH that is moved to another RP to be deleted if identified by moid (`--id`). In this case parentResourcepool would previously have returned an error causing this to bail and not delete the VCH even though it is found by its moid

hickeng replied:
`d.parentResourcePool` can have the following values:
a. moid of originally targeted resource pool (endpointVM may not still be in that pool)
b. current parent pool of the endpointVM

In either case we now delete children of that pool that can be identified as cVMs, then attempt to destroy that pool if it's empty.

I'm unsure what happens if the endpointVM is the last VM in the cluster and the pool returned is the root pool for the cluster. I do not know what happens if you call destroy on the root resource pool of a cluster, but to be safe I'd rather we never do so.

andrewtchin commented on lib/install/management/appliance.go:
```diff
@@ -93,6 +94,25 @@ func (d *Dispatcher) isVCH(vm *vm.VirtualMachine) (bool, error) {
+		err = errors.Errorf("Failed to fetch guest info of appliance vm: %s", err)
+		return false, err
+	}
+	extraconfig.Decode(extraconfig.MapSource(info), &cspec)
```
fixed

andrewtchin commented on lib/install/management/delete.go:
```diff
@@ -46,10 +46,28 @@ func (d *Dispatcher) DeleteVCH(conf *config.VirtualContainerHostConfigSpec) erro
+		log.Warnf("No container VMs found, but proceeding with delete of VCH due to --force")
+		err = nil
+	}
+	if d.parentResourcepool != nil {
```
this is required since I use d.parentResourcepool above and that could be nil if the parent RP was deleted OOB

hickeng commented on lib/install/management/delete.go:
```diff
@@ -46,10 +46,28 @@ func (d *Dispatcher) DeleteVCH(conf *config.VirtualContainerHostConfigSpec) erro
+			return err
+		}
+		// Assume the cVMs and RP have already been deleted
+		log.Warnf("No container VMs found, but proceeding with delete of VCH due to --force")
```
I do not see a path by which both `err` and `d.parentResourcepool` are non-nil so this logic will simply skip the attempt to delete the VCH when we hit the `d.parentResourcePool != nil` condition.

The text about containerVMs is also misleading given we've not looked for containerVMs at this point.

hickeng commented on tests/test-cases/Group6-VIC-Machine/6-03-Delete.md:
```diff
@@ -36,3 +36,13 @@ This test requires that a vSphere server is running and available
+5. Delete the VM and RP to cleanup
+
+### Expected Outcome:
+1. All steps should succeed
```
We need another two tests:
1. endpointVM is moved out of it's pool, pool is left alone
2. endpointVM is moved into the root of the cluster and the original pool is deleted.

andrewtchin replied:
Just to be clear, for 1) the endpointVM should be successfully deleted and the RP and its children should not be deleted
and for 2) the original pool should be deleted OOB after endpointVM is moved to the root pool then vic-machine delete should successfully delete the endpointVM?
Thanks

hickeng replied:
1) correct
2) correct - and other VMs in the root pool should be left alone (which is what the current test asserts but worth being explicit)

andrewtchin commented:
Added the integration tests, working through them now.
@hickeng 
- For "I think we should remove the failover path for the actual parent pool - this code was added in #3116 but I'm unsure what specific scenario was being addressed with this path." what failover path is that that i need to remove?
- I have confirmed that on create we always create a new RP so we can't have a VCH in the root RP. Given that do we still need this? "a sanity check to return nil if the ComputeResource[0] is the root pool."
- Does this "I do not know what happens if you call destroy on the root resource pool of a cluster, but to be safe I'd rather we never do so." mean that after we have deleted a VCH that was moved into the root RP we do not want to call the pool destroy?

hickeng commented:
* This is the failover path I'm talking about - https://github.com/vmware/vic/pull/3116/files#diff-786c272aaac433e0eaae7299e08b52b7R91
* If we remove the failover path that switches to the actual parent and not the recorded parent then no, we do not need the sanity check.
* Correct - we do not want to even try to destroy the root pool

There should be a follow up issue to look at disabling _Move_ for the individual VMs.

hickeng commented on lib/install/management/delete.go:
```diff
@@ -46,10 +46,28 @@ func (d *Dispatcher) DeleteVCH(conf *config.VirtualContainerHostConfigSpec) erro
+		}
+		// Can't find the RP VCH was created in to delete cVMs, continue anyway
+		log.Warnf("No container VMs found, but proceeding with delete of VCH due to --force")
+		err = nil
```
getComputeResource never returns non-nil for both rp and err. If you set `err = nil` here then we will still skip over the delete of the containerVMs (not sure why the method is called `DeleteVCHInstances`).

I don't know what that will mean for deleting images if they are still attached to cVMs - I suspect it will fail with errors about locked vmdks if any of the cVMs are powered on. I don't see that that will cause any other "interesting" behaviours like deleting non-VCH related things, but please sanity check me if you haven't looked at this path.

hickeng commented on tests/test-cases/Group6-VIC-Machine/6-03-Delete.robot:
```diff
@@ -95,3 +104,128 @@ Attach Disks and Delete VCH
+
+    # Delete with force
+    ${moid}=  Get VM Moid  %{VCH-NAME}
+    ${ret}  ${output}=  Run And Return Rc And Output  bin/vic-machine-linux delete --target %{TEST_URL} --user %{TEST_USERNAME} --password=%{TEST_PASSWORD} --compute-resource=%{TEST_RESOURCE} --id ${moid} --force
```
As a note, if you're specifying `--id` then I don't think you need `--compute-resource`. I frequently use `--id` to override which VCH I'm deleting when working with nimbus or vmc.

hickeng commented on tests/test-cases/Group6-VIC-Machine/6-03-Delete.robot:
```diff
@@ -95,3 +104,128 @@ Attach Disks and Delete VCH
+
+
+Delete VCH moved from its RP
+    Run Keyword If  '%{HOST_TYPE}' == 'ESXi'  Pass Execution  Test skipped on ESX due to unable to move into RP
```
ESX does have resource pools and we should be able to move VMs between them. It's possible that govc is missing exposure of the appropriate command.
(I'm not sure how to view resource pools in the H5 client however).

andrewtchin commented:
Replacement PR in #6816 to get the build to run
